op-bindings/bindgen/etherscan_contracts.go

```diff
@@ -0,0 +1,142 @@
+	etherscanGetDeploymentTxHashUrl = "https://api.etherscan.io/api?module=contract&action=getcontractcreation&contractaddresses=%s&apikey=%s "
+	etherscanGetDeployedBytecodeUrl = "https://api.etherscan.io/api?module=proxy&action=eth_getCode&address=%s&tag=latest&apikey=%s"
+	etherscanGetTxByHashUrl         = "https://api.etherscan.io/api?module=proxy&action=eth_getTransactionByHash&txHash=%s&tag=latest&apikey=%s"
+)
+
+// readEtherscanContractsList reads a JSON file specified by the given file path and
+// parses it into a slice of `etherscanContract`.
+//
+// Parameters:
+// - filePath: The path to the JSON file containing the list of etherscan contracts.
+//
+// Returns:
+// - A slice of etherscanContract parsed from the JSON file.
+// - An error if reading the file or parsing the JSON failed.
+func readEtherscanContractsList(filePath string) (etherscanContracts, error) {
+	var data contractsData
+	err := readJSONFile(filePath, &data)
+	return data.Etherscan, err
+}
+
+// genEtherscanBindings generates Go bindings for Ethereum smart contracts based on the ABI and bytecode
+// fetched from Etherscan.
+// The function reads the list of contracts from the provided file path and fetches the ABI and
+// bytecode for each contract from Etherscan using the provided API key. It then generates Go bindings
+// for each contract and writes metadata for each contract to the specified output directory.
+//
+// Parameters:
+// - contractListFilePath: Path to the file containing the list of contracts.
+// - sourceMapsListStr: Comma-separated list of source maps.
+// - etherscanApiKey: API key to fetch data from Etherscan.
+// - goPackageName: Name of the Go package for the generated bindings.
+// - metadataOutputDir: Directory to output the generated contract metadata.
+//
+// Returns:
+//   - An error if there are issues reading the contract list, fetching data from Etherscan, generating
+//     contract bindings, or writing contract metadata.
+func genEtherscanBindings(contractListFilePath, etherscanApiKey, goPackageName, metadataOutputDir string, apiMaxRetries, apiRetryDelay int) error {
+	contracts, err := readEtherscanContractsList(contractListFilePath)
+	if err != nil {
+		return fmt.Errorf("error reading contract list %s: %w", contractListFilePath, err)
+	}
+
+	// TODO - Reimplement
+	// if len(contracts.Verified) == 0 && len(contracts.Unverified) == 0 {
+	// 	return fmt.Errorf("No contracts parsable from given contract list: %s", contractListFilePath)
+	// }
+
+	tempArtifactsDir, err := mkTempArtifactsDir()
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err := os.RemoveAll(tempArtifactsDir)
+		if err != nil {
+			log.Printf("Error removing temporary directory %s: %v", tempArtifactsDir, err)
+		} else {
+			log.Printf("Successfully removed temporary directory")
+		}
+	}()
+
+	contractMetadataFileTemplate := template.Must(template.New("contractMetadata").Parse(contractMetadataTemplate))
+
```

Comment:
These functions are nearly the same, it makes me think there can be a better way to do this

Reply:
Refactored